python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1107 @@
+        # Register services on client
+        rpc.register_service(KVSTORE_PULL, 
+                             PullRequest, 
+                             PullResponse)
+        rpc.register_service(KVSTORE_PUSH, 
+                             PushRequest, 
+                             None)
+        rpc.register_service(INIT_DATA, 
+                             InitDataRequest, 
+                             InitDataResponse)
+        rpc.register_service(BARRIER, 
+                             BarrierRequest, 
+                             BarrierResponse)
+        rpc.register_service(REGISTER_PUSH,
+                             RegisterPushHandlerRequest, 
+                             RegisterPushHandlerResponse)
+        rpc.register_service(REGISTER_PULL, 
+                             RegisterPullHandlerRequest, 
+                             RegisterPullHandlerResponse)
+        rpc.register_service(GET_SHARED,
+                             GetSharedDataRequest,
+                             GetSharedDataResponse)
+        rpc.register_service(GET_FULL_SHAPE,
+                             GetFullShapeRequest,
+                             GetFullShapeResponse)
+        # Store the tensor data with specified data name
+        self._data_store = {}
+        # Store the full data shape across kvserver
+        self._full_data_shape = {}
+        # Basic information
+        self._server_namebook = rpc.read_ip_config(ip_config)
+        self._server_count = len(self._server_namebook)
+        self._group_count = self._server_namebook[0][3]
+        self._machine_count = int(self._server_count / self._group_count)
+        self._client_id = rpc.get_rank()
+        self._machine_id = rpc.get_machine_id()
+        self._part_id = self._machine_id
+        self._main_server_id = self._machine_id * self._group_count
+        # TODO(chao) : remove tmp file using new shared-tensor API
+        self._open_file_list = []
+        # push and pull handler
+        self._pull_handler = self._default_pull_handler
+        self._push_handler = self._default_push_handler
+        random.seed(time.time())
+
+    def __del__(self):
+        """Finalize KVClient
+        """
+        # TODO(chao) : remove tmp file using new shared-tensor API
+        for file in self._open_file_list:
+            if(os.path.exists(file)):
+                os.remove(file)
+
+    def register_push_handler(self, func):
+        """Register UDF push function on server.
+
+        client_0 will send this request to all servers.
+
+        Parameters
+        ----------
+        func : UDF push function
+        """
+        if self._client_id == 0:
+            request = RegisterPushHandlerRequest(func)
+            # send request
+            for server_id in range(self._server_count):
+                rpc.send_request(server_id, request)
+            # wait confirmation
+            for _ in range(self._server_count):
+                response = rpc.recv_response()
```

Comment:
the func isn't invoked in the client?

Reply:
You are right, it's a bug. I changed it.